### ISSUE TITLE
False negative of prop-types with stateless components

### ISSUE THREAD

**Issue by kossnocorp** [yannickcr/eslint-plugin-react#735]
When a stateless component is created as a property of an object, `prop-types` gives false negative:

🚫 reports "'children' is missing in props validation":

```jsx
Panel.Body = ({children}) => (
  <div className='Panel-body'>
    {children}
  </div>
)

Panel.Body.propTypes = {
  children: PropTypes.node
}
```

👌🏽 works fine:

```jsx
const Body = ({children}) => (
  <div className='Panel-body'>
    {children}
  </div>
)

Body.propTypes = {
  children: PropTypes.node
}

Panel.Body = Body
```

**Comment by kentcdodds** [comment]
Another example:

🚫 reports error:

```javascript
function Foo({bar}) {
  return <div>{bar}</div>
}

Foo.propTypes = {
  bar: PropTypes.string.isRequired
}
```

Passes:

```javascript
Foo.propTypes = {
  bar: PropTypes.string.isRequired
}

function Foo({bar}) {
  return <div>{bar}</div>
}
```

**Comment by yannickcr** [comment]
@kossnocorp How is created your `Panel` object ? To match the propTypes with `Panel.Body` we're searching for the `Panel` variable in the current scope but in your example `Panel` is never declared so we cannot makes the match.

This code is working:
```jsx
var Panel = {}; // <= Declare Panel
Panel.Body = ({children}) => (
  <div className='Panel-body'>
    {children}
  </div>
)

Panel.Body.propTypes = {
  children: PropTypes.node
}
```

@kentcdodds What version of eslint and eslint-plugin-react are you using? Your first example is working for me with eslint-plugin-react `v6.1.2`

**Comment by kossnocorp** [comment]
@yannickcr the `Panel` is declared as a class exported by default, maybe this is the problem. However, it's valid ES200X and supposed to be working 😇

I have errors:

![tmux___users_koss_src_kossnocorp_make](https://cloud.githubusercontent.com/assets/52201/17837795/716b47e4-67c5-11e6-8d16-0c41aae5cba5.jpg)

... on such code:

```jsx
import React, {PropTypes, Component} from 'react'
import classNames from 'classnames'

export default class Panel extends Component {
  static propTypes = {
    children: PropTypes.node,
    color: PropTypes.oneOf(['green', 'red', 'orange', 'grey', 'blue'])
  }

  render () {
    const {color, children} = this.props
    const panelClasses = classNames('Panel', {
      [`is-${color}`]: color,
      'has-color': color
    })
    return (
      <section className={panelClasses}>
        {children}
      </section>
    )
  }
}

Panel.Header = ({children}) => (
  <header className='Panel-header'>
    {children}
  </header>
)

Panel.Header.propTypes = {
  children: PropTypes.node
}

Panel.Body = ({children}) => (
  <div className='Panel-body'>
    {children}
  </div>
)

Panel.Body.propTypes = {
  children: PropTypes.node
}
```

**Comment by kossnocorp** [comment]
@kentcdodds I just tried your code and I don't have the error:

![tmux___users_koss_src_kossnocorp_make](https://cloud.githubusercontent.com/assets/52201/17837841/b354caf8-67c6-11e6-94c5-0f718fee082e.jpg)

You can try to update I suppose.

**Comment by yannickcr** [comment]
> the `Panel` is declared as a class

Ok, seems the issue comes from here. I'll have a look.

**Comment by kentcdodds** [comment]
Huh, it looks like my issue is possibly related but I'm not sure what's going on... Here's how I can reproduce the issue:

```javascript
import React, {PropTypes} from 'react'

function Foo({bar}) { // react/prop-types error reported for bar on this line
  const {baz} = Foo
  return <div>{baz} {bar}</div>
}

Foo.propTypes = {
  bar: PropTypes.string.isRequired,
}

Foo.baz = 'hi'
```

This does not produce any error:

```javascript
import React, {PropTypes} from 'react'

function Foo({bar}) {
  const baz = Foo.baz
  return <div>{baz} {bar}</div>
}

Foo.propTypes = {
  bar: PropTypes.string.isRequired,
}

Foo.baz = 'hi'
```

[Here's a real use case](https://github.com/javascriptair/site/blob/30b5930230bf8cb98ce21c04d21afac72d0bd612/src/pages/episode/sections/show-notes/person-notes.js#L70-L100)

Looks like destructuring a property off of the function (whether inside the function itself or not) is causing issues.

**Comment by yannickcr** [comment]
@kentcdodds ok, it's a different problem then. 

Can you open a new issue with your last message? Thanks.

**Comment by kentcdodds** [comment]
Cool, will do, thanks!
